### PR TITLE
Fixed Asset Depreciation

### PR DESCRIPTION
**Pull request by akshay83**
As discussed in the discussion Thread https://discuss.frappe.io/t/fixed-asset-depreciation/4308/6

### CODE REVIEW

File: erpnext/accounts/doctype/fixed_asset_account/depreciation_report.py [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+import frappe
+from datetime import date, timedelta
+from datetime import datetime
+from erpnext.accounts.utils import get_fiscal_year
+from frappe.utils import flt
+
+def getDateDiffDays(d1, d2):
+    return abs((d2 - d1).days)
+    
+def getDeprTillLastYear(day_before_start, fixed_asset_name):
```
Comment (rmehta) [inline erpnext/accounts/doctype/fixed_asset_account/depreciation_report.py]: use the Pythonic casing e.g. `get_depr_till_last_year`

File: erpnext/accounts/doctype/fixed_asset_account/depreciation_report.py [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+        for accdepr in frappe.get_doc("Fixed Asset Account", fixed_asset_name).depreciation:
+           if get_fiscal_year(fiscal_year = accdepr.fiscal_year) == get_fiscal_year(date = day_before_start):
+               return accdepr.total_accumulated_depreciation
+        return 0
+        
+def getDeprProvidedThisYear(method, depronopening, purchase_value, deprtilllastyr, rateofdepr):
```
Comment (rmehta) [inline erpnext/accounts/doctype/fixed_asset_account/depreciation_report.py]: its very hard for someone to read `depronopening` should be `depreciation_at_opening`

File: erpnext/accounts/doctype/fixed_asset_account/depreciation_report.py [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+    finyrfrom = datetime.strptime(finyrfrom, "%Y-%m-%d").date()
+    finyrto = datetime.strptime(finyrto, "%Y-%m-%d").date()
+    day_before_start = finyrfrom - timedelta (days=1)    
+    TOTAL_DAYS_IN_YEAR = 365
+    method = frappe.get_doc("Company", company).default_depreciation_method
+    global ps
+    if fa_name==None:
+	    ps = frappe.db.sql("""select led.*,ifnull((select sum(sale.asset_purchase_cost) from `tabFixed Asset Sale` sale where sale.fixed_asset_account=led.fixed_asset_name and sale.docstatus=1  and sale.posting_date>=%s and sale.posting_date<=%s),0) as total_sale_value
+           from `tabFixed Asset Account` led where is_sold=false or (is_sold=true and (select count(*) from `tabFixed Asset Sale` sale where sale.fixed_asset_account=led.fixed_asset_name and docstatus=1 and sale.posting_date>=%s and sale.posting_date<=%s)>0)""", (finyrfrom,finyrto,finyrfrom,finyrto), as_dict=True)
+    else:
```
Comment (rmehta) [inline erpnext/accounts/doctype/fixed_asset_account/depreciation_report.py]: don't cross 80 cols per line, break the query into a new line

File: erpnext/accounts/doctype/fixed_asset_account/depreciation_report.py [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+           from `tabFixed Asset Account` led where (is_sold=false or (is_sold=true and (select count(*) from `tabFixed Asset Sale` sale where sale.fixed_asset_account=led.fixed_asset_name and docstatus=1 and sale.posting_date>=%s and sale.posting_date<=%s)>0)) and led.fixed_asset_name=%s""", (finyrfrom,finyrto,finyrfrom,finyrto,fa_name), as_dict=True)
+
+
+    for assets in ps:
+        fixed_asset_name = assets.fixed_asset_name
+        fixed_asset_account = assets.fixed_asset_account
+        rateofdepr = abs(assets.depreciation_rate)
+        purchase_date = datetime.strptime(assets.purchase_date, "%Y-%m-%d").date()
+
+        global deprtilllastyr # Depreciation provided till close of last fiscal Yr.
```
Comment (rmehta) [inline erpnext/accounts/doctype/fixed_asset_account/depreciation_report.py]: very bad design to use globals, wrap it in a class if you want to share variables

**Comment by rmehta** [comment]
@akshay83 some common style issues:

1. Naming is very important - read this https://signalvnoise.com/posts/3250-clarity-over-brevity-in-variable-and-method-names
2. Make small functions (if your function crosses 10 lines, break it) - it becomes very hard to maintain code if its not readable
3. Make your sql queries cleaner to read - use multi-line strings.

Also please see some inline comments.

**Comment by akshay83** [comment]
Tried to make the changes as suggested.

File: erpnext/accounts/doctype/fixed_asset_account/fixed_asset_account.js~ [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+		method: 'post_journal_entry',
+		doc: frm.doc,
+		callback: function(r) {
+			frm.fields_dict.post_journal_entry.$input.addClass("btn-primary");
+			var doclist = frappe.model.sync(r.message);
+			frappe.set_route("Form", doclist[0].doctype, doclist[0].name);
+		}
+	});
+});
+
```
Comment (rmehta) [inline erpnext/accounts/doctype/fixed_asset_account/fixed_asset_account.js~]: You are committing temp files. Add `.gitignore` for `js~` files

**Comment by rmehta** [comment]
@akshay83 Code looks great now.

Final step is adding automated tests. This will ensure that the feature does not break in future releases. You might even find a bug or two if you are lucky.

This will help us in understanding the use-cases too.

See this link:

https://frappe.io/docs/user-guide/writing-tests

Just a simple example

```
class TestAssetAccount(unittest.TestCase):
  def test_purchase(self):
    account = frappe.get_doc({
      "doctype": "Fixed Asset Account",
      "name": "xxx"
      ... })
    account.insert()
    je_name = account.post_journal_entry()
    
    je = frappe.get_doc("Journal Entry", je_name)
    self.assertTrue("Fixed Assets" in [d.account for d in je.accounts])
```

File: erpnext/accounts/doctype/fixed_asset_account/depreciation_report.py [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+from datetime import datetime
+from erpnext.accounts.utils import get_fiscal_year
+from frappe.utils import flt
+
+def get_date_difference_in_days(d1, d2):
+    return abs((d2 - d1).days)
+
+class Depreciation:
+	def __init__(self,financial_year_from, financial_year_to, company, fixed_asset=None):
+	    self.financial_year_from = datetime.strptime(financial_year_from, "%Y-%m-%d").date()
```
Comment (rmehta) [inline erpnext/accounts/doctype/fixed_asset_account/depreciation_report.py]: keep consistent comma + space 

```
def __init__(self, financial_year_from, financial_year_to
                  ^
```

**Comment by akshay83** [comment]
I am trying to test the "Tests". However I am getting this : 

```
erpnext@erpnext-dev:~/frappe-bench/apps/erpnext$ bench frappe --run_tests -d "Fixed Asset Account"
Traceback (most recent call last):
  File "/home/erpnext/frappe-bench/env/bin/frappe", line 9, in <module>
    load_entry_point('frappe==4.9.3', 'console_scripts', 'frappe')()
  File "/home/erpnext/frappe-bench/apps/frappe/frappe/cli.py", line 54, in main
    return run(fn, parsed_args)
  File "/home/erpnext/frappe-bench/apps/frappe/frappe/cli.py", line 84, in run
    out = globals().get(fn)(**args)
  File "/home/erpnext/frappe-bench/apps/frappe/frappe/cli.py", line 68, in new_fn
    return fn(*args, **new_kwargs)
  File "/home/erpnext/frappe-bench/apps/frappe/frappe/cli.py", line 824, in run_tests
    tests=tests, force=force)
  File "/home/erpnext/frappe-bench/apps/frappe/frappe/test_runner.py", line 21, in main
    raise Exception, 'db_name must start with "test_"'
Exception: db_name must start with "test_"
```

I dont have any Idea...Pls help

**Comment by rmehta** [comment]
Create a new site starting with `test_` e.g. `test_site`

This is to stop you from accidentally deleting data on other sites.

**Comment by akshay83** [comment]
Ummm :-( ... I don't seem to be getting this right

```
erpnext@erpnext-dev:~/frappe-bench/sites$ bench frappe --run_tests -d "Fixed Asset Account" test_site
Traceback (most recent call last):
  File "/home/erpnext/frappe-bench/env/bin/frappe", line 9, in <module>
    load_entry_point('frappe==4.9.3', 'console_scripts', 'frappe')()
  File "/home/erpnext/frappe-bench/apps/frappe/frappe/cli.py", line 54, in main
    return run(fn, parsed_args)
  File "/home/erpnext/frappe-bench/apps/frappe/frappe/cli.py", line 84, in run
    out = globals().get(fn)(**args)
  File "/home/erpnext/frappe-bench/apps/frappe/frappe/cli.py", line 68, in new_fn
    return fn(*args, **new_kwargs)
  File "/home/erpnext/frappe-bench/apps/frappe/frappe/cli.py", line 824, in run_tests
    tests=tests, force=force)
  File "/home/erpnext/frappe-bench/apps/frappe/frappe/test_runner.py", line 32, in main
    ret = run_tests_for_doctype(doctype, verbose=verbose, tests=tests, force=force)
  File "/home/erpnext/frappe-bench/apps/frappe/frappe/test_runner.py", line 72, in run_tests_for_doctype
    make_test_records(doctype, verbose=verbose, force=force)
  File "/home/erpnext/frappe-bench/apps/frappe/frappe/test_runner.py", line 127, in make_test_records
    make_test_records(options, verbose, force)
  File "/home/erpnext/frappe-bench/apps/frappe/frappe/test_runner.py", line 127, in make_test_records
    make_test_records(options, verbose, force)
  File "/home/erpnext/frappe-bench/apps/frappe/frappe/test_runner.py", line 127, in make_test_records
    make_test_records(options, verbose, force)
  File "/home/erpnext/frappe-bench/apps/frappe/frappe/test_runner.py", line 128, in make_test_records
    make_test_records_for_doctype(options, verbose, force)
  File "/home/erpnext/frappe-bench/apps/frappe/frappe/test_runner.py", line 173, in make_test_records_for_doctype
    frappe.local.test_objects[doctype] += make_test_objects(doctype, test_module.test_records, verbose)
  File "/home/erpnext/frappe-bench/apps/frappe/frappe/test_runner.py", line 200, in make_test_objects
    d = frappe.copy_doc(doc)
  File "/home/erpnext/frappe-bench/apps/frappe/frappe/__init__.py", line 720, in copy_doc
    newdoc = get_doc(copy.deepcopy(d))
  File "/home/erpnext/frappe-bench/apps/frappe/frappe/__init__.py", line 440, in get_doc
    return frappe.model.document.get_doc(arg1, arg2)
  File "/home/erpnext/frappe-bench/apps/frappe/frappe/model/document.py", line 44, in get_doc
    controller = get_controller(doctype)
  File "/home/erpnext/frappe-bench/apps/frappe/frappe/model/document.py", line 64, in get_controller
    module = load_doctype_module(doctype, module_name)
  File "/home/erpnext/frappe-bench/apps/frappe/frappe/modules/__init__.py", line 60, in load_doctype_module
    doctype_python_modules[key] = frappe.get_module(get_module_name(doctype, module, prefix))
  File "/home/erpnext/frappe-bench/apps/frappe/frappe/__init__.py", line 482, in get_module
    return importlib.import_module(modulename)
  File "/usr/lib/python2.7/importlib/__init__.py", line 37, in import_module
    __import__(name)
ImportError: No module named budget_distribution.budget_distribution
```

**Comment by rmehta** [comment]
Have your rebased your repo? Budget distribution is now renamed. This should ideally work.

@neilLasrado can you help?

**Comment by neilLasrado** [comment]
Try doing a git pull --rebase upstream v5.0
and the run bench frappe --latest
also remember to pull the latest in your frappe repo also.
Do let me know if you have any trouble getting this done.

**Comment by akshay83** [comment]
I ran the commands, but now the desk the empty.I can only see the top menu with command bar.

**Comment by ankitjavalkarwork** [comment]
run `bench frappe -b`

Why don’t you join in on IRC at #frappe

**Comment by akshay83** [comment]
No success, still getting the same error "No module named budget_distribution.budget_distribution". tried IRC Too but could not get through

**Comment by neilLasrado** [comment]
when do you get this error exactly ?
Did bench frappe --latest work without any errors ?
and are you on the latest v5 version on your local frappe repo.

**Comment by neilLasrado** [comment]
I think im getting this ...
try frappe bench --latest test_site
let me know if any error arises

**Comment by neilLasrado** [comment]
Please ping me on the IRC ... I am online...

**Comment by nabinhait** [comment]
May be "budget_distribution" directory still exists in your local filesystem.

If you do not have any untracked file in your local erpnext repo, then run `git clean -df`, it will remove all the unwanted files and directories.

You can manually remove the directory as well.

**Comment by akshay83** [comment]
Sorry Guys. Its still not working. giving the same error No module named budget_distribution.budget_distribution. I have even tried it on the clean demo VM provided with no customization.

As per our previous discussion the step I was skipping was installing the erpnext app in the test_site. that's why i was able to run bench frappe --run_tests test_site. But as soon as I install the erpnext app, the tests breaks. "ALL Tests"

Could this be related to the "migration" switch in bench. To reproduce use the demo VM dated 15 Sep 2014.
* bench update
* ->bench-repo$ git pull
* bench update
* bench migrate-to-v5
* bench update
* bench new-site test_site
* bench frappe --install_app erpnext test_site

**Comment by rmehta** [comment]
Added a new feature and test today and it worked fine for me.

Maybe your branch is not rebased properly

https://github.com/frappe/erpnext/commit/5bceebc6a1a763bca957a3cdc3dabfc4124a2ed6#diff-fcd53c77c1b9c153bcd27e98bc1dffddL6

**Comment by nabinhait** [comment]
I have pushed a fix, can you please update and try again?

**Comment by akshay83** [comment]
now getting this http://pastebin.com/DkQsUy3Z

**Comment by nabinhait** [comment]
pushed the fix

**Comment by akshay83** [comment]
Opened #2635

**Pull request closed without merging**